docs/testing-guidelines.md

```diff
@@ -0,0 +1,71 @@
+# Testing Guidelines for UDS Marketplace Apps
+
+## Introduction
+
+This document outlines the testing guidelines and best practices for UDS Marketplace Apps. These guidelines aim to ensure consistency, reliability, and quality across all UDS packages.
+
+## 1. Journey Testing
+
+Journey testing is crucial for ensuring that UDS packages function correctly within their deployment environment and integrate properly with other UDS components.
+
+#### **Definition**
+A **journey** in the context of UDS package testing is defined as:
+> The minimal and least fragile way to test all features of an application that could be broken from how we deploy it.
+
+#### **Key Principles**
+- Focus on testing the deployment, not the application itself (that's the upstream developer's job).
+- Ensure that our deployment method doesn't break key functionalities.
+- Test integration with UDS core components (e.g., Istio, Keycloak).
+
+#### **Implementation Guidelines**
+- Design tests that cover essential functionality without being overly complex.
+- Focus on deployment-related issues (e.g., network policies, SSO access, cluster resource access).
+- Handle limitations due to licensing or other constraints by documenting them and implementing what testing is possible.
+
+#### **Tools**
+- UI Testing: Playwright, Cypress
+- Non-UI Testing: Jest
+- Custom Scripts: Bash or other scripting languages as needed
+
+## 2. Upgrade Testing
+Ensure that the current development package works when deployed over the previously released version.
+
+#### **Key Considerations**
+- Test data migration and persistence.
+- Verify that configurations are properly carried over or updated.
+- Check for any breaking changes in APIs or integrations.
+
+## 3. Linting and Static Analysis
+
+Use appropriate tooling to ensure code quality and consistency.
+
+#### **Recommended Tools**
+- YAML Linting: `yamllint`
+- Zarf Package Linting: `zarf dev lint`
+
+## **Best Practices**
+
+1. **Consistency**: Use similar testing approaches across different UDS packages where possible.
+2. **Documentation**: Clearly document what each test is checking and why.
+3. **Maintenance**: Keep tests updated as the application and deployment methods evolve.
+4. **CI/CD Integration**: Incorporate all types of tests into your CI/CD pipeline.
+5. **Error Handling**: Ensure tests fail gracefully and provide clear error messages.
+6. **Idempotency**: Design tests to be repeatable and produce consistent results.
+7. **Performance Consideration**: While thorough, tests should not unnecessarily slow down the development process.
```

Comment:
```suggestion
7. **Performance Consideration**: While thorough, tests should not unnecessarily slow down the development process - beyond 30 minutes is too long for a single package test.
```

Reply:
We could tighten that but from my experience 30 mins is _way_ too long for CI of a single package.

Reply:
I would love to say 15 min because that's generally my maximum threshold for CI to run but it is possible that this shape of test requires more - thoughts?

Reply:
From my experience (external from here) long tests make for slow feedback loops (unless you can emulate behavior locally). I'd be happy to reduce this as much as you think is permissible. Anecdotally when I ran through the tests on `uds-core` the other day it took >20 mins for at least one of the tests there, so I just want to make sure it aligns with expectations.

Reply:
While i agree philosophically with @daveworth < 15 min is not really possible if "test" is defined as the "full CI job to deploy said app and then any tests executed against it running to completion", rather than "after package is deployed run a suite of tests". 

It takes >15 min to deploy and test a gitlab install and ~18 min to run the upgrade and then test. 

The full SWF integration repo bundle test (bundle not package on full uds-core) takes just over 30 min, but is testing like 6 packages so its a little different. 

I think less than 30 min is a good threshold for a rule. That being said its the upper limit. Most package tests should be sub 10 min.

Reply:
I did my best to encapsulate the above as part of this commit: https://github.com/defenseunicorns/uds-common/pull/205/commits/18853a49170b3cb54ba43d0f6f58d7fb6d9accf1